discord_bot/discord_bot.py

```diff
@@ -1,60 +1,85 @@
+import asyncio
 import json
 import requests
-import asyncio
 import websockets
 
+from .opcodes import Opcodes
+
 CONFIG_FILE_PATH = 'configs/config.json'
-DISCORD_GATEWAY_ENDPOINT = 'https://discordapp.com/api/v6/gateway'
 
 
 class DiscordBot:
+    """
+    Establishes a connection to the discord gateway and handles varied messages
+    """
 
     def __init__(self):
-
         # instance variables
         self.config = self.load_config()
         self.gateway_ws_url = self.get_gateway()
+        self.heartbeat_interval_ms = None
+        self.last_seq = None
+        self.event_loop = asyncio.get_event_loop()
+        self.websocket = None
 
-        # actions below
-        asyncio.get_event_loop().run_until_complete(self.gateway_handler())
+        # actions
+        self.event_loop.run_until_complete(self.gateway_handler())
 
     @staticmethod
     def load_config():
+        """
+        Loads the configurations
+        :return: dict - the configurations
+        """
         with open(CONFIG_FILE_PATH) as f:
             return json.load(f)
 
-    @staticmethod
-    def get_gateway():
-        r = requests.get(DISCORD_GATEWAY_ENDPOINT)
+    def get_gateway(self):
+        """
+        Caches a gateway value and retrieves a new URL
+        :return: gateway URL
+        """
+        r = requests.get(self.config("discord_gateway_endpoint"))
         return r.json()['url']
 
-    async def handshake(self, websocket):
-        identity_obj = {
-            "token": self.config["bot_token"],
-            "properties": {
-                "$os": "windows",
-                "$browser": "disco",
-                "$device": "disco"
-            },
-            "compress": False,
-            "large_threshold": 250,
-            "shard": [1, 10]
-        }
-        await websocket.send(json.dumps({"op": 2, "d": identity_obj}))
+    async def send_json(self, payload):
+        asyncio.ensure_future(self.websocket.send(json.dumps(payload)))
 
-    async def gateway_handler(self):
+    async def handshake(self, message):
+        """
+        When a websocket connection is opened, Hello payload is received. Then, an Identify payload is sent as
+        part of the handshake to authorize this client.
 
+        :param message: dict - hello payload
+        """
+        self.heartbeat_interval_ms = message["d"]["heartbeat_interval"]
+        asyncio.ensure_future(self.heartbeat())
+
+        handshake_identity = self.config["handshake_identity"]
+        asyncio.ensure_future(self.send_json({"op": Opcodes.IDENTIFY, "d": handshake_identity}))
+
+    async def heartbeat(self):
+        """
+        Sends a heartbeat payload every heartbeat interval.
+        """
+        asyncio.ensure_future(asyncio.sleep(self.heartbeat_interval_ms / 1000.0))
```

Comment:
This won't work, this will create a separate task that will sleep while the program continues to run without waiting for the sleep. We need to wait for this to be done before continuing.